### PR TITLE
En 7380 early end of epoch

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added early end of epoch implementation that can be used in the hardfork process
- changed how bech32PubkeyConverter does the stack dump
- added interceptor & resolver debugger in hardfork exporter

### CODE REVIEW

File: integrationTests/consensus/testInitializer.go [hunk 1]
```diff
@@ -9,6 +9,7 @@ import (
 	"time"
 
 	"github.com/ElrondNetwork/elrond-go/crypto/peerSignatureHandler"
+	"github.com/ElrondNetwork/elrond-go/epochStart/notifier"
```
Comment (BeniaminDrasovean) [inline integrationTests/consensus/testInitializer.go:13]: Remove empty line.

File: scripts/testnet/hardfork.sh [hunk 2]
```diff
@@ -7,7 +7,12 @@ source "$ELRONDTESTNETSCRIPTSDIR/include/config.sh"
 VALIDATOR_RES_PORT="$PORT_ORIGIN_VALIDATOR_REST"
 
 if [ -z "$1" ]; then
-  echo "epoch argument was not provided. Usage: './hardfork.sh [epoch number]' as in './hardfork.sh 1'"
+  echo "epoch argument was not provided. Usage: './hardfork.sh [epoch number] [forced: true|false]' as in './hardfork.sh 1 false'"
+  exit
+fi
+
+if [ -z "$2" ]; then
+  echo "epoch argument was not provided. Usage: './hardfork.sh [epoch: number] [forced: true|false]' as in './hardfork.sh 1 false'"
```
Comment (bogdan-rosianu) [inline scripts/testnet/hardfork.sh:15]: `forced` argument, not epoch

File: api/hardfork/routes_test.go [hunk 3]
```diff
@@ -99,7 +99,7 @@ func TestTrigger_TriggerCanNotExecuteShouldErr(t *testing.T) {
 
 	expectedErr := errors.New("expected error")
 	ws := startNodeServer(&mock.HardforkFacade{
-		TriggerCalled: func(epoch uint32) error {
+		TriggerCalled: func(epoch uint32, forced bool) error {
```
Comment (bogdan-rosianu) [inline api/hardfork/routes_test.go:102]: `forced` not used

File: api/hardfork/routes_test.go [hunk 4]
```diff
@@ -144,7 +144,7 @@ func TestTrigger_ManualShouldWork(t *testing.T) {
 	}
 	buffHr, _ := json.Marshal(hr)
 	ws := startNodeServer(&mock.HardforkFacade{
-		TriggerCalled: func(epoch uint32) error {
+		TriggerCalled: func(epoch uint32, forced bool) error {
```
Comment (bogdan-rosianu) [inline api/hardfork/routes_test.go:147]: `forced` not used

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: api/hardfork/routes.go [hunk 5]
```diff
@@ -18,14 +18,15 @@ const (
 
 // FacadeHandler interface defines methods that can be used by the gin webserver
 type FacadeHandler interface {
-	Trigger(epoch uint32) error
+	Trigger(epoch uint32, forced bool) error
```
Comment (sasurobert) [inline api/hardfork/routes.go:21]: withEarlyEndOfEpoch

File: epochStart/interface.go [hunk 6]
```diff
@@ -2,18 +2,18 @@ package epochStart
...
 	"github.com/ElrondNetwork/elrond-go/data"
 	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/data/state"
 )
 
 // TriggerHandler defines the functionalities for an start of epoch trigger
 type TriggerHandler interface {
 	Close() error
-	ForceEpochStart(round uint64) error
+	ForceEpochStart()
```
Comment (sasurobert) [inline epochStart/interface.go:16]: Don't you need current round?

File: update/trigger/trigger.go [hunk 7]
```diff
@@ -202,7 +205,16 @@ func (t *trigger) computeAndSetTrigger(epoch uint32, originalPayload []byte) (bo
 		t.recordedTriggerMessage = originalPayload
 	}
 
-	if epoch > t.epochProvider.MetaEpoch() {
+	if forced {
```
Comment (sasurobert) [inline update/trigger/trigger.go:208]: With early end of eppch

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: epochStart/interface.go (hunk 6 above)
Comment (iulianpascalau) [reply epochStart/interface.go:16]: nope, it is taken automatically as the next possible round. Took into consideration the `MinRoundsBetweenEpochs` config value

File: integrationTests/consensus/testInitializer.go (hunk 1 above)
Comment (iulianpascalau) [reply integrationTests/consensus/testInitializer.go:13]: done

File: scripts/testnet/hardfork.sh (hunk 2 above)
Comment (iulianpascalau) [reply scripts/testnet/hardfork.sh:15]: done

File: api/hardfork/routes_test.go (hunk 3 above)
Comment (iulianpascalau) [reply api/hardfork/routes_test.go:102]: done

File: api/hardfork/routes_test.go (hunk 4 above)
Comment (iulianpascalau) [reply api/hardfork/routes_test.go:147]: done

File: api/hardfork/routes.go (hunk 5 above)
Comment (iulianpascalau) [reply api/hardfork/routes.go:21]: done

File: update/trigger/trigger.go (hunk 7 above)
Comment (iulianpascalau) [reply update/trigger/trigger.go:208]: done

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Pull request merged**